xbmc/cores/AudioEngine/AESinkFactory.cpp

```diff
@@ -38,41 +38,55 @@ bool CAESinkFactory::HasSinks()
   return !m_AESinkRegEntry.empty();
 }
 
-void CAESinkFactory::ParseDevice(std::string &device, std::string &driver)
+AESinkDevice CAESinkFactory::ParseDevice(const std::string& device)
 {
-  int pos = device.find_first_of(':');
+  AESinkDevice dev{};
+  dev.name = device;
+
+  size_t pos = dev.name.find_first_of(':');
   bool found = false;
-  if (pos > 0)
+
+  if (pos != std::string::npos)
   {
-    driver = device.substr(0, pos);
+    dev.driver = device.substr(0, pos);
 
     for (const auto& reg : m_AESinkRegEntry)
     {
-      if (!StringUtils::EqualsNoCase(driver, reg.second.sinkName))
+      if (!StringUtils::EqualsNoCase(dev.driver, reg.second.sinkName))
         continue;
 
-      device = device.substr(pos + 1, device.length() - pos - 1);
+      dev.name = dev.name.substr(pos + 1, dev.name.length() - pos - 1);
       found = true;
     }
   }
 
   if (!found)
-    driver.clear();
+    dev.driver.clear();
```

Comment:
probably best to return here instead of continuing.
Otherwise when device = "foo:bar" and foo is not a registered driver, we'd end up with driver = "", name = "foo" and friendly name = "bar".

Reply:
Yes, for better compatibility is supposed driver is optional. Then these cases are now handled properly:

`{38C5D27A-785F-4088-8DBF-4FCBB40DC813}:HDMI - PA249 (NVIDIA High Definition Audio)`

`{38C5D27A-785F-4088-8DBF-4FCBB40DC813}`

For simplicity friendly name is discarted as anyway is "fixed" later in `ValidateOutputDevices`

```
driver = ""
device = {38C5D27A-785F-4088-8DBF-4FCBB40DC813}
friendlyName = ""
```

Reply:
Tried disabling the WASAPI driver and indeed the displayed name in the settings screen doesn't have the Directsound: prefix anymore, but guisettings.xml still has it => not a situation with no driver name.
Back compatibility is good but this may be excessive if there is no real use case. I grabbed old version as far as available on https://mirrors.kodi.tv (2015 - v15) to find out.
A fixed format driver:name:friendly name would be easier to work with.

Reply:
> A fixed format driver:name:friendly name would be easier to work with.

The check for the possible lack of driver was already in the existent code and it is not complicated to maintain this.

```
  if (!found)
    driver.clear();
```

I honestly don't see any problem here.

Reply:
Blank driver is not a real use case and would be better treated as an error, but fine go ahead it's a detail.